include/nbl/system/demote_promote_writer_readers_lock.h

```diff
@@ -0,0 +1,217 @@
+	* @brief Acquires lock for reading. This is the operation with the lowest priority, thread will be blocked until there are no writers and no readers waiting for a writer upgrade
+	*/
+	void read_lock()
+	{
+		constexpr uint32_t preemptedMask = StateSemantics{ .currentReaders = 0,.pendingWriters = MaxActors,.pendingUpgrades = MaxActors,.writing = true,.stateLocked = false };
+		for (uint32_t spinCount = 0; true; spinCount++)
+		{
+			uint32_t actual = state.fetch_or(flipLock);
+			// stateLocked: some other thread is doing work and expects to update state
+			if (actual & flipLock)
+			{
+				if (spinCount > SpinsBeforeYield)
+					std::this_thread::yield();
+				continue;
+			}
+			// If there are any pending writers, upgrades or a thread is currently writing, cannot acquire lock (`preemptedMask`)
+			// Must release flipLock and yield for other threads to progress
+			if (actual & preemptedMask)
+			{
+				state.store(actual);
+				if (spinCount > SpinsBeforeYield)
+					std::this_thread::yield();
+				continue;
+			}
+			actual += static_cast<uint32_t>(StateSemantics{ .currentReaders = 1 });
+			// release the flipLock, increment currentReaders 
+			state.store(actual);
+			reading = true;
+			break;
+		}
+	}
+
+	/**
+	* @brief Release lock after reading. 
+	*/
+	void read_unlock()
+	{
+		for (uint32_t spinCount = 0; true; spinCount++)
+		{
+			uint32_t actual = state.fetch_or(flipLock);
+			// stateLocked: some other thread is doing work and expects to update state
+			if (actual & flipLock)
+			{
+				if (spinCount > SpinsBeforeYield)
+					std::this_thread::yield();
+				continue;
+			}
+			actual -= static_cast<uint32_t>(StateSemantics{ .currentReaders = 1 });
+			// release the flipLock, decrement currentReaders
+			state.store(actual);
+			break;
+		}
+	}
+
+	/**
+	* @brief Acquires lock for writing. Is superseded by reader upgrades: if any thread with a reader's lock is trying to upgrade, then this thread will be blocked until that reader becomes a writer
+	*/
+	void write_lock()
+	{
+		constexpr uint32_t preemptedMask = StateSemantics{ .currentReaders = MaxActors,.pendingWriters = 0,.pendingUpgrades = MaxActors,.writing = true,.stateLocked = false };
+		bool registeredPending = false;
+		for (uint32_t spinCount = 0; true; spinCount++)
+		{
+			uint32_t actual = state.fetch_or(flipLock);
+			// stateLocked: some other thread is doing work and expects to update state
+			if (actual & flipLock)
+			{
+				if (spinCount > SpinsBeforeYield)
+					std::this_thread::yield();
+				continue;
+			}
+			// If there are any pending upgrades or a thread is currently reading or writing, cannot acquire lock (`preemptedMask`), so release flipLock and yield for other threads to progress
+			// Also registers this thread as a pending writer if not done so already
+			if (actual & preemptedMask)
+			{
+				if (!registeredPending)
+				{
+					registeredPending = true;
+					actual += static_cast<uint32_t>(StateSemantics{ .pendingWriters = 1 });
+				}
```

Comment:
you want to register as a pending writer ALWAYS no matter if you're getting preempted or not (how else will everyone know that you're pending), its just the state lock you need to acquire